lib/rdoc/parser/prism_ruby.rb

```diff
@@ -0,0 +1,1081 @@
+          @scanner.process_comments_until(node.location.end_line)
+        end
+      else
+        @scanner.skip_comments_until(node.location.end_line)
+      end
+    end
+
+    def visit_def_node(node)
+      start_line = node.location.start_line
+      end_line = node.location.end_line
+      @scanner.process_comments_until(start_line - 1)
+
+      case node.receiver
+      when Prism::NilNode, Prism::TrueNode, Prism::FalseNode
+        visibility = :public
+        singleton = false
+        receiver_name = { Prism::NilNode => 'NilClass', Prism::TrueNode => 'TrueClass', Prism::FalseNode => 'FalseClass' }[node.receiver.class]
+        receiver_fallback_type = :class
+      when Prism::SelfNode
+        # singleton method of a singleton class is not documentable
+        return if @scanner.singleton
+        visibility = :public
+        singleton = true
+      when Prism::ConstantReadNode, Prism::ConstantPathNode
+        visibility = :public
+        singleton = true
+        receiver_name = constant_path_string(node.receiver)
+        receiver_fallback_type = :module
+        return unless receiver_name
+      when nil
+        visibility = @scanner.visibility
+        singleton = @scanner.singleton
+      else
+        # `def (unknown expression).method_name` is not documentable
+        return
+      end
+      name = node.name.to_s
+      signature = MethodSignature.new(node)
+      tokens = @scanner.visible_tokens_from_location(node.location)
+
+      @scanner.add_method(
+        name,
+        receiver_name: receiver_name,
+        receiver_fallback_type: receiver_fallback_type,
+        visibility: visibility,
+        singleton: singleton,
+        signature: signature,
+        tokens: tokens,
+        start_line: start_line,
+        end_line: end_line
+      )
+    ensure
+      @scanner.skip_comments_until(end_line)
+    end
+
+    def visit_constant_path_write_node(node)
+      @scanner.process_comments_until(node.location.start_line - 1)
+      path = constant_path_string(node.target)
+      return unless path
+
+      @scanner.add_constant(
+        path,
+        constant_path_string(node.value) || node.value.slice,
+        node.location.start_line,
+        node.location.end_line
+      )
+      @scanner.skip_comments_until(node.location.end_line)
+      # Do not traverse rhs not to document `A::B = Struct.new{def undocumentable_method; end}`
+    end
+
+    def visit_constant_write_node(node)
+      @scanner.process_comments_until(node.location.start_line - 1)
+      @scanner.add_constant(
+        node.name.to_s,
+        constant_path_string(node.value) || node.value.slice,
+        node.location.start_line,
+        node.location.end_line
+      )
+      @scanner.skip_comments_until(node.location.end_line)
+      # Do not traverse rhs not to document `A = Struct.new{def undocumentable_method; end}`
+    end
+
+    private
+
+    def constant_arguments_names(call_node)
+      return unless call_node.arguments
```

Comment:
Let's move these argument attribute's presence check to the caller so readers can see the code flow more easily.

Reply:
I think we should leave this.
`Prism::CallNode#arguments` returns nil when there is no argument. If it returns an empty argument node, it was simple but it's not. I think this is to make the footprint small. Maybe related to https://github.com/ruby/prism/issues/1915

In my opinion, caller side should not know about the detail of CallNode structure.
If it is a call_node, caller should just call `constant_arguments_names` to check if call_node's argument is a list of constants.

Added some test for this in https://github.com/ruby/rdoc/pull/1144/commits/65317242a1089171047c03d4b6ef4b2a30a1e432, Found documenting code that includes `alias_method 42 if false` fails with error in master